google/cloud/storage/benchmarks/async_throughput_benchmark.cc

```diff
@@ -0,0 +1,298 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/port_platform.h"
+#if GOOGLE_CLOUD_CPP_STORAGE_HAVE_GRPC && GOOGLE_CLOUD_CPP_HAVE_COROUTINES
+#include "google/cloud/storage/async_client.h"
+#include "google/cloud/storage/benchmarks/benchmark_utils.h"
+#include "google/cloud/options.h"
+#include "google/cloud/status.h"
+#include "google/cloud/testing_util/command_line_parsing.h"
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace {
+
+auto constexpr kDescription = R"""(
+This program benchmarks uploads to GCS (Google Cloud Storage) and downloads from
+it using the google::cloud::storage_experimental::AsyncClient APIs.
+
+The benchmark tries to answer the following questions:
+
+1) How much CPU is used by the client library for uploads?
+2) How much CPU is used by the client library for downloads?
+
+The answer to these questions depends on many parameters, including the size of
```

Comment:
"The answers to these questions depend on ..."

Reply:
Done